rest/config.go

```diff
@@ -513,65 +513,40 @@ func (clusterConfig *ClusterConfig) GetCredentials() (string, string, string) {
 	return base.TransformBucketCredentials(clusterConfig.Username, clusterConfig.Password, *clusterConfig.Bucket)
 }
 
-// Reads a ServerConfig from raw data
-func ReadServerConfigFromData(runMode SyncGatewayRunMode, data []byte) (*ServerConfig, error) {
+// LoadServerConfig loads a ServerConfig from either a JSON file or from a URL
+func LoadServerConfig(runMode SyncGatewayRunMode, path string) (config *ServerConfig, err error) {
+	var data io.ReadCloser
 
-	data = base.ConvertBackQuotedStrings(data)
-	var config *ServerConfig
-	if err := json.Unmarshal(data, &config); err != nil {
-		return nil, err
-	}
-
-	config.RunMode = runMode
-
-	// Validation:
-	if err := config.setupAndValidateDatabases(); err != nil {
-		return nil, err
-	}
-
-	return config, nil
-}
-
-// Reads a ServerConfig from a URL.
-func ReadServerConfigFromUrl(runMode SyncGatewayRunMode, url string) (*ServerConfig, error) {
-
-	resp, err := http.Get(url)
-	if err != nil {
-		return nil, err
-	}
-	defer resp.Body.Close()
-	responseBody, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return nil, err
-	}
-	return ReadServerConfigFromData(runMode, responseBody)
-
-}
-
-// Reads a ServerConfig from either a JSON file or from a URL.
-func ReadServerConfig(runMode SyncGatewayRunMode, path string) (*ServerConfig, error) {
 	if strings.HasPrefix(path, "http://") || strings.HasPrefix(path, "https://") {
-		return ReadServerConfigFromUrl(runMode, path)
+		resp, err := http.Get(path)
+		if err != nil {
+			return nil, err
+		} else if resp.StatusCode >= 300 {
+			return nil, base.HTTPErrorf(resp.StatusCode, http.StatusText(resp.StatusCode))
+		}
+		data = resp.Body
 	} else {
-		return ReadServerConfigFromFile(runMode, path)
+		data, err = os.Open(path)
+		if err != nil {
+			return nil, err
+		}
 	}
+
+	defer data.Close()
+	return readServerConfig(runMode, data)
 }
 
-// Reads a ServerConfig from a JSON file.
-func ReadServerConfigFromFile(runMode SyncGatewayRunMode, path string) (*ServerConfig, error) {
-	file, err := os.Open(path)
-	if err != nil {
-		return nil, err
-	}
-	defer file.Close()
+// readServerConfig returns a validated ServerConfig from an io.Reader
+func readServerConfig(runMode SyncGatewayRunMode, r io.Reader) (config *ServerConfig, err error) {
```

Comment:
I like the idea of readServerConfig taking an io.Reader, instead of duplicating the code in 'FromData' and 'FromFile'.  

However, it feels a bit obfuscating for ConvertBackQuotedStrings to accept a reader, do a readAll, then return a different reader (via bytes.NewBuffer).  

What do you think about a 'DecodeServerConfig' helper that accepts an io.Reader, does the readAll/defer close, calls convertBackQuotedStringsBytes, decodes, and returns (ServerConfig, err).  That could then be called from here, and also from getDbConfigFromServer.

Reply:
> However, it feels a bit obfuscating for ConvertBackQuotedStrings to accept a reader, do a readAll, then return a different reader (via bytes.NewBuffer).

Agreed, will change that back to dealing purely with bytes.

> What do you think about a 'DecodeServerConfig' helper that accepts an io.Reader, does the readAll/defer close, calls convertBackQuotedStringsBytes, decodes, and returns (ServerConfig, err). That could then be called from here, and also from getDbConfigFromServer.

`getDbConfigFromServer` is doing things a bit differently, as it's marshalling a dbConfig rather than a ServerConfig. I've created `decodeAndSanitiseConfig` to do the converting and decoding for both types.

I've kept the io.Reader rather than io.ReadCloser to make testing a bit easier - (`bytes.Buffer` doesn't implement Closer)